src/useFormState.ts

```diff
@@ -128,6 +137,80 @@ export function useFormState<T, I>(opts: UseFormStateOpts<T, I>): ObjectState<T>
   return form;
 }
 
+export type ObjectStateCache<T> = Record<string, [ObjectState<T>, T]>;
+type UseFormStatesOpts<T> = {
+  config: ObjectConfig<T>;
+  autoSave?: (state: ObjectState<T>) => Promise<void>;
+  getId: (t: T) => string;
+};
+type GetObjectStateProps<T, I> = {
+  readOnly?: boolean;
```

Comment:
Fwiw I'd personally probably purposefully drop `readOnly` for the mvp of `useFormStates` and work on adding it back later; but your call if it's easy to do now anyway.